### DISCUSSION TITLE
NEEY 4A Balancer

### DISCUSSION THREAD

**Discussion by kmn13** [shining-man/bsc_fw#12] [General]
Ich starte hier einmal eine Runde bezüglich der NEEY Balancer. Zwar gibt es schon ein paar Kommentare unter anderen Kapiteln - denke jedoch eine "Kanalisierung" macht Sinn.

@shining-man soweit ich gelesen habe, hast du die Anbindung mit 1x NEEY im Einsatz mit dem BSC.
Zur Prüfung der Performance etc, kann ich gerne etwas beitragen.
Ich habe bei mir 8 Stück "herumliegen" und diese sollen später dann kaskadiert eingesetzt werden.

Leider scheitere ich aktuell mit meinem Testboard ([https://amzn.eu/d/dFsBpOJ]) noch etwas mit einer generellen Verbindung.
Benutzte _BSC Version 0.1.5_ und einen aktuellen NEEY 4A der Generation 4 (von Aliexpress - HankzorBMS).
Der NEEY will einfach nicht gefunden werden. Wobei mit der "GeeWe BMS" App oder auch einem BLE Sniffer klappt die Verdindung.
Natürlich nicht gleichzeitig, aber das "blaue" Lämpchen zeigt an ob eine Verbindung besteht oder nicht.

Meine NEEYs sind: Hardware version: HW-2.8.0 und Software version: ZH-1.2.3

@shining-man welche Version vom NEEY hast du erfolgreich im Einsatz?

**Comment by shining-man** [comment]
Richtig, ich habe einen NEEY im Einsatz. FW/HW-Version schaue ich später einmal.

Sobald du die Seite Schnittstellen->Bluetooth offen hast, wird nach BT Devices gescannt und am unteren Ende der Seite angezeigt. Hier sollte er, unabhängig der Version des NEEY, die gefundenen Devices auflisten. 
Wiki: https://github.com/shining-man/bsc_fw/wiki#schnittstellen---bluetooth
Was zeigt er hier bei dir an? Listet er überhaupt ein Device auf?

**Reply by kmn13** [reply]
Grundsätzlich hat er schon Devices aufgelistet^^

Kann aber nun mal Entwarnung geben. Scheint so, als wäre _mein ESP32 Board Schrott_ 👎 konnte nun mal einen anderen D1 Mini herauszaubern und siehe da, **Geäte sofort gefunden und auch connected.**

Werde die nächsten Tage einmal alle anbinden und versuchen die Performace zu vergleichen ;-)

**Reply by shining-man** [reply]
Ich bin gespannt ob das Verbinden von mehreren Devices problemlos klappt. Konnte ich leider bisher noch nicht testen. Auch wie dann der RAM verbrauch bzw. der freie Heap ausschaut. In Verbindung mit dem Redesign kommt dann ein neues Relais. Hier habe ich noch einmal ca. 20k Heap frei machen können.
Wäre super wenn du mich auf dem laufenden hältst:-)

Im Wiki habe ich eine List mit den unterstützten/funktionierenden Geräte inkl. FW/HW Version angefangen.

**Reply by kmn13** [reply]
Mahlzeit ;-)
Mittlerweile konnte ich meine **Tests erfolgreich abschließen** 😎 und nach einer kleinen Erweiterung in den Arrays auch 7x NEEYs per BLE verbinden. 
In meinem aktuellen Testaufbau - der sich vorerst nur auf WLAN Verbindung mit dem Heimnetzwerk, 7x NEEY und Timeout-Alarmen je NEEY beschränkt - läuft es auch über längere Zeit (getestet >2h) gleich stabil wie mit 1x. 👍
Allerdings bricht auch die Verbidung mit 1x NEEY nach einer gewissen Zeit ab und kommt dann wieder.

Verwendete Version: **BSC V0.1.5** und ESP32 Board von KeyStudio [https://amzn.eu/aocP1zk](https://amzn.eu/aocP1zk)

**Anmerkungen**
Über einen kleinen 🐞 im `BmsData.ccp` bin ich noch gestolpert: `bmsLastDataMillis` hat eine falsche Arraylänge -> korigiert auf `unsigned long bmsLastDataMillis[BT_DEVICES_COUNT+SERIAL_BMS_DEVICES_COUNT];`

Die entsprechenden Konstanten in `Defindes.h`, `nimconfig.h` und `params.h` habe ich von 5 auf 7 erweitert.

Nach dem Verbinden aller NEEYs blieb der Heap im Großen und Ganzen stabil (wie erwartet)..ca 2k Unterschied zwischen 0 und 7 NEEY.
### 
Ich habe mehrere ESP32 Boards getetstet (_alle mit 240MHz_):

1. KeyStudio ESP32 KS0413 (das einzige mit original Espressif Chip und kommt so den eingesetzen am ähnlichsten):
Flotte Verbindung, auch nach dem Verbinden Webservice bedienbar. 👍
2. AZDelivery ESP32 D1 Mini
  verbindet zu allen NEEYs, hat jedoch merklich schlechtere Performance im Webservice. Bis zu hängern über mehrer Sekunden
3. AZDelivery ESP32 Dev Kit C V4
    findet alle BT Geräte, kackt jedoch während dem Verbinden ab und rebootet 👎
4. SP-Cow ESP32
  findet alle BT Geräte, kackt jedoch während dem Verbinden ab und rebootet 👎
5. AZDelivery ESP32 Dev Kit C V2
  mein erstes Board, vermutlich defekt, da nicht einmal die BT Suche korrekt funktionierte 👎

Zu prüfen und berücksichtigen:
- die NEEY ballern mit 500ms ihre Information über BLE

Ein paar allgemeine "wünsch-dir-was" Funktionen sind mit auch noch eingefallen - sind aber wahrscheinlich in den "Issues" besser aufgehoben?

**Reply by shining-man** [reply]
Das hört sich dich erst einmal gut an. Ich habe fast schon erwartet, dass es an der ein oder anderen Stelle klemmt. Ich übernehme die Erweiterung auf 7 BT Devices, dann in die nächste Version.
Der Fehler bei `unsigned long bmsLastDataMillis[BT_DEVICES_COUNT+SERIAL_BMS_DEVICES_COUNT];` hatte ich auch schon entdeckt. Aber danke fürs melden👍.

Was meinst du mit:
```
Zu prüfen und berücksichtigen:
-    die NEEY ballern mit 500ms ihre Information über BLE
```
Sollte eigentlich kein Problem sein. Er muss ja "nur" die Daten wegkopieren.
Ausgewertet werden sie erst im Alarmrules Task, der alle Sekunde läuft.
Mit einem NEEY, einem seriellen BMS und CAN gibt es jedenfalls keinen Probleme. Da hatte ich mir das Timing angeschaut.

Wann verliert er denn die Verbindung zu dem NEEY. Passiert das zyklisch? Das ist mir bisher noch nicht aufgefallen. Ich muss wohl einmal nach meiner FW Version vom NEEY schauen. Ist nur nicht so schön, da ich dann die Verbindung zum BSC trennen muss. Taucht in der Debugausgabe eine Meldung auf?
Als ich damals die Kommunikation zum NEEY gesnifft habe, hat die App einaml ein Anfrage gesendet. Daraufhin sendet der NEEY kontinuierlich seine Daten. Ich weiß nicht ob es evtl. sinn macht die Anfrage zyklische alle xx Minuten zusenden, damit es zu keinem Abbruch kommt? Könnte ich zum Testen einmal einbauen. 

Wenn ich den aktuellen Stand pushe, könntest du den evtl. mit den NEEY's testen?
Hat sich ein bisschen was geändert. RAM sparen über alles ;-)

Das mit den "zweifelhaften" ESPs ist einer der Gründe warum ich kein NodeMCU oder sonstiges auf meiner PCB wollte. Das gibt, dann nur Ärger, wenn es bei jemandem nicht funktioniert. Das Ding muss 24/7 laufen.

Wünsche einfach in die Issues👍.

**Reply by kmn13** [reply]
Meinte mit

> Zu prüfen und berücksichtigen:
> -    die NEEY ballern mit 500ms ihre Information über BLE

eine Idee / Spur zu finden, warum es eventuell die BT Verbindung wieder killt. 

Leider bekomme ich keine "beschreibende" Logausgabe über die Terminal Verbindung. Ich sehe nur das `onDisconnect` aus dem BleHandler.
Die Logs aus der NimBLE lib bringe ich noch nicht raus.

Zeitlich ist es leider ganz verschieden.
Vorhin bei 7 verbundenen 2h lange nicht. Aber später bei nur 1 Verbindung nach 37min, dann 50min und jetzt läufts seit 130min.

Ich kann gerne deinen Letztstand mal reinballern und schauen was so los ist 😎 vielleicht löst sich dadurch auch das Problem (wobei ich trotzdem den Grund noch herausfinden möchte...)

> Das mit den "zweifelhaften" ESPs ist einer der Gründe warum ich kein NodeMCU oder sonstiges auf meiner PCB wollte. Das gibt, dann nur Ärger, wenn es bei jemandem nicht funktioniert. Das Ding muss 24/7 laufen.

Da bin ich ganz deiner Meinung, ich mag da auch keine Überaschungseier.

**Reply by shining-man** [reply]
Mit folgenden Build flags solltes du eigentlich auch die NIMBLE Logs bekommen:
build_flags =   -DCORE_DEBUG_LEVEL=5
                      -DLOG_LOCAL_LEVEL=ESP_LOG_DEBUG

Wenn du den letzten Stand von mir nutzt dann würde ich in der define.h
#define DEBUG_ON_FS  auskommentieren und
#define DEBUG_ON_HW_SERIAL  einfügen


Disconnecten alle sieben gleichzeitig, oder immer mal wieder einer?

**Reply by kmn13** [reply]
Hab gestern Abend noch den Letztstand gezogen und etwas herumprobiert.
Die V.0.2.1 macht echt nen gewaltigen Sprung nach vorne 👍

Beim Disconnecten ist mir aufgefallen, dass es keinen Unterschied macht, ob ich 1 NEEY oder 7 NEEYs aktiviert habe.
Aber im Fall von 7 verhält sich der "Disconnect" wie eine Kettenreaktion. Es sterben alle 7 Verbindungen nacheinander 😥

Ich füge einmal 2 Logs (gekürzt aus den wesentlichen Zeitraum) hier an:
(Bei einer Suche nach "`Disconnect`" kommt man schnell zu Stelle)
* [BSC_1_of_2NEYY_ShortlyConnectedThenDisconnectAndBLENeverOn-221203-210827.log](https://github.com/shining-man/bsc_fw/files/10148561/BSC_1_of_2NEYY_ShortlyConnectedThenDisconnectAndBLENeverOn-221203-210827.log)
Hier habe ich 2 NEEY konfiguriert, jedoch wird nur 1x gefunden und auch verbunden. Nach einer Weile wird es Disconnected (leider konnte ich noch keine Ursache finden) und möchte sich nicht mehr verbinden, wird auch nicht gefunden.
Unterschied zum nachfolgenden LOG, es waren noch nicht alle BuildFlags on, darum sind nicht alle "NimBLE" Flags angezeigt, leider...
Muss mal was einbauen um zu sehen, ob der BLE Tast sich irgendwie verabschiedet und nicht mehr kommt ? kann das das möglich sein?
(_kurzer Beisatz: ist mein erstes Projekt mit ESP32, darum fehlt mir mit dieser CPU etwas die Erfahrung_ 😅)

* [BSC_7NEEY_Disconnect-221203-201840.log](https://github.com/shining-man/bsc_fw/files/10148562/BSC_7NEEY_Disconnect-221203-201840.log)
Die Verbindung wird Disconnected und nacheinander folgen alle, jedoch wird wieder automatisch verbunden.
Die Einträge mit "`() NimBLE: Controller not ready to receive packets`" kann ich noch nicht ganz zuordnen ob relevant, treten aber auch auf, wenn nur 1x NEEY Verbunden ist (fehlen nur im ersten Log wegen der BuildFlags)

Bei den obigen beiden Logs, war auch immer eine Verbindung zum Webservice aktiv. Mein WLAN hätte ich einmal als gut eingestuft. Luftlinie 3m zum Router. 

Über Nacht habe ich dann noch folgende Tests laufen lassen:
1x ESP mit 2x NEEY
1x ESP mit 5x NEEY
jeweils kein Webservice Zugriff aktiv.
Am Morgen waren alles NEEY Verbunden, wobei das Log zeigte das es Abbrüche gab, die Verbindung jedoch wieder aufgebaut werden konnten (Dauer der Neuverbindung ca 700ms, laut Log - daher auch kein Fehler).

Habe jetzt einmal Testweise im `BleHandler `die `pBLEScan->setMaxResults(10);`, das könnte für mich die Ursache sein, dass gelegentlich, wenn viele unverbundene BLE Devices im Raum sind, die Liste nich all meine NEEY oder zumindest das gesuchte NEEY enthält.

Für die Ursache der Abrisse, fehlt mir noch der Plan 😅

**Reply by shining-man** [reply]
Mit kommt das alles ein bisschen komisch vor. Bei mir läuft das seit Wochen/Monaten ohne einen einzigen Verbindungsabbruch.
Mein NEEY hat die Version HW-2.8.0/SW-1.2.1. Dies SW Version unterscheidet sich also ein bisschen von deiner.

Ich würde mir erst einmal die Stromversorgung vom ESP anschauen. Hast du ein Oszi um zu sehen, wie die Spannung im Falle eines Disconnect ausschaut. Wenn nicht, dann würde ich einfach einmal einen dicken Elko direkt an die +3V3 vom WROOM Modul machen.

Die Logs werde ich mir einmal genauer anschauen, ob ich Auffälligkeiten finde.

Die Tasks werden jetzt schon auf Laufzeit überwacht. Wenn der BT Task nicht merh laufen sollte müsstest du dies auf der Weboberfläche in der ersten Kachel (System) sehen. Hier wird dann in rot die Tasknummer angezeigt die nicht merh antwortet. Im Log müsste dann ein Eintrag TaskRunSate mit der fehlerhaften Tasknummer kommen.

**Reply by shining-man** [reply]
Ich habe am Reconnect etwas geändert, denn dieser sollte ja auf jeden fall funktionieren. Probier mal den letzten Stand aus dem bt_test Branch aus.
Probier es erst einmal mit einem Device. Ich möchte sehen ob hier auch schon `NimBLE: Controller not ready to receive packets` Meldungen kommen. Solange ein Device nicht sauber funktioniert, werden es mehrere erst recht nicht.
Selber kann ich es nicht testen, da es bei mir zu keinen Problemen kommt.

**Reply by kmn13** [reply]
Hab beim Chinesen einmal die Unterschiede in der SW angefragt ;-) mal sehen ob eine Antwort kommt^^

Die Stromversorgung hätte ich als gut eingeschätzt, aber Oszi mit Trigger hängt mal in der +3V3 Leitung drinnen ;-)

Den Letztstand aus dem bt_test hab ich schon raufgeballert...

Folgendes Verhalten bezüglich _Controller not ready.._
1xNEEY: keine Meldungen im Log
2xNEEY: Meldung kommt: `NimBLE: Controller not ready to receive packets (cmd)`

...aktuell logge ich mal das System mit 1xNEEY ... es sollte eigentlich so stabil sein/bleiben wie bei dir ;-)

**Reply by kmn13** [reply]
Habe das System die ganze Nacht durchlaufen lassen mit 1xNEEY
**- ohne Probleme, ohne Disconnect.**
Also so wie erwartet und wie es auch bei dir läuft 👍

**Reply by shining-man** [reply]
Probiere es einmal mit zwei Geräten und ändere vorher im BleHandler in der onConnect-Funktion die Zeile `pClient->updateConnParams(120,120,0,150);` in  `pClient->updateConnParams(24,240,3,500);`

**Reply by shining-man** [reply]
Wenn es nicht funktioniert bitte wieder einen Log

**Reply by shining-man** [reply]
Wenn es nicht funktioniert könntest du in der nimconfig.h `CONFIG_BT_NIMBLE_LOG_LEVEL` auf 0 setzen, dann wird mehr geloggt.

**Reply by kmn13** [reply]
alles klar ;-)
Ich hab die Änderungen einmal reingestopft - aber leider selbes Spiel 😢

Folgende Log konnte ich erstellen:
CONFIG_BT_NIMBLE_LOG_LEVEL = 0
[NIMBLE_LOGLEVEL_0-221205-210413.log](https://github.com/shining-man/bsc_fw/files/10157592/NIMBLE_LOGLEVEL_0-221205-210413.log)
Das verkraftet meine CPU nicht und geht zwischen dem Verbinden vom 1. und 2. NEEY in reboot.
Habe auch testweise den TAG Alarm deaktiviert, brachte ab eh nichts (wollte ein wenig die Serielle entlasten).

CONFIG_BT_NIMBLE_LOG_LEVEL = 1
[NIMBLE_LOGLEVEL_1-221205-205448.log](https://github.com/shining-man/bsc_fw/files/10157601/NIMBLE_LOGLEVEL_1-221205-205448.log)
Habe einmal versucht etwas weniger Loginfo. Stürzt nicht mehr ab, aber natürlich keine so genauen Daten

**Reply by kmn13** [reply]
🥳 ich habe es nun geschafft, die Verbindungen zu mehreren NEEYs ohne diese doofen `NimBLE: Controller not ready to receive packets` Meldungen zu erhalten ;-) 
Es laufen nun auch wieder alle 7 und keine verdächtige Meldung^^

Nach nun längeren probieren und suchen bin ich über folgende Info gestolpert: [https://github.com/h2zero/NimBLE-Arduino/blob/release/1.4/docs/Usage_tips.md#put-ble-functions-in-a-task-running-on-the-nimble-stack-core](url)

Nun wird ja im nimconfig.h die NimBLE auf _Core 1_ gepinnt:  `#define CONFIG_BT_NIMBLE_PINNED_TO_CORE 1`
Bis jetzt dachte ich, der ganze Appcode läuft auch am _Core 1_ -> bei mir läuft die `task_ble `jedoch am _Core 0_.
Festgestellt mit:
```
void task_ble(void *param)
{
  ESP_LOGD(TAG, "-> 'task_ble' runs on core %d", xPortGetCoreID());
.....
```

Ich habe nun die `task_ble `einmal auf denselben Core wie die NimBLE gepinnt:   `xTaskCreatePinnedToCore(task_ble,"ble", 3000, NULL, 5, &task_handle_ble, CONFIG_BT_NIMBLE_PINNED_TO_CORE);`

Beim pinnen von Task und Lib auf Core 1 kommen die Fehlermeldungen, mit **Core 0 nicht** mehr👍.

Da ich sonst noch keine andere Hardware laufen habe, kann ich nicht feststellen ob nun andere Nebeneffekte hinzugekommen sind.
Aktuell sieht es einmal stabil aus - lief jetzt mind. 6h ohne _Disconnects_ oder _Controller not ready to..._ Meldungen durch 👍

@shining-man gab es einen speziellen Grund, dass die  `#define CONFIG_BT_NIMBLE_PINNED_TO_CORE 1` von default: auskommentiert (=0) auf 1 umgebogen wurde? Nur um zu wissen, wo vielleicht nun andere Nebengeräusche auftauchen könnten ;-)

**Reply by shining-man** [reply]
Ich habe den NimBLE auf Core 1 gepinnt, da ich sporadisch immer wieder Reboots aufgrund des Watchdogs hatte. Der NimBLE hat den Core 0 immer wieder blockiert und es somit zum reboot kam. Die meisten reboots  sind bereits beim Connect des NEEYS aufgetreten. Das würde ich zumindest einmal testen. Ich würde mir eine kleine Testbench machen und z.B. den NEEY über ein Relais immer wieder aus-/einschalten und schauen ob er sauber verbindet. Und ich würde den Anlauf und den anschließenden ersten Connect anschauen und auch dies ausgiebig testen.

Vielleicht sollte man alle anderen Task fest auf Core 1 pinnen. Ich bin der Meinung, dass die Tasks bei meinen Tests auch auf Core 1 liefen und ich den BT Task dann eben auch auf Task 1 gepinnt habe...

**Reply by shining-man** [reply]
Ich habe mir noch einmal einen alten SW-Stand angeschaut, mit dem ich damals die Probleme hatte.
Mit der alten Version kann ich die damiligen Probleme reprodzieren. Mit der Version kommen dann wieder die Probleme mit den Reboots. Die Reboots treten meistens beim Starten des ersten BT-Scans nach dem Start des BSC auf. Festpinnen des NimBLE auf Core 1 bringt sofort Abhilfe. Ich wollte jetzt nicht mehr zuviel Zeit in die alte Version investieren. Mir ist nur aufgefallen, dass in der alten Version der BT-Task auch auf Core 1 läuft. In der aktuellen SW-Version läuft der BT-Task auf Core 0. Alle anderen Tasks laufen in der aktuellen Version aber auf Task 1. Woher das kommt, ist mir nicht ganz klar.

Ich habe jetzt in der aktuellen Version den BT-Task und NimBLE auf Core 0 gepinnt und alle anderen Tasks auf Core 1. So läuft es jetzt schon seit mehr als 24 h ohne Reboot. Parallel laufen einige Testroutinen in der die meisten Schnittstellen (3x Serial, CAN, onwire, Ansteuerung der DIs/DOs) ziemlich stark ausgelastet werden. 

Was ich zwischen den Versionen grundlegend geändert habe, ist die Version der espressif32-platform und der NimBLE Lib. In der NimBLE Lib V1.4.0 wurde auch der NimBLE -Core aktualisiert. 
Meine Vermutung ist, dass es irgendwie daran liegt. Da nach dem Tausch bei mir alles weiterhin funktioniert hat, habe ich auch keine weiteren Untersuchungen gemacht.

Ich habe die Tasks jetzt auf die entsprechenden Cores festgepinnt, damit es nicht zu „selbstständigen“ Änderungen bezüglich der Core Zuordnung kommt.

Ich pushe noch meinen aktuellen Teststand. Wenn du ihn auch noch einmal mit deinen 7 NEEYs testen kannst, wäre super👍, denn dann würde ich die Änderung in den Masterbranch übernehmen.

**Comment by kmn13** [comment]
Ich war beruflich unterwegs und konnte daher nicht weiter testen.

Werde mir den Letztstand ziehen und mal testen. Morgen Abend starte ich dann auch noch die Verbindungsversuche mit zyklischem aus/ein.
Gebe dir dann Feedback, wenn ich durch bin 😎

**Reply by kmn13** [reply]
Sodale 😅 habe jetzt mal einiges getestet:

* Langzeit Test (seit gestern Abend) mit 7xNEEY: lief durch ohne Probleme, jedoch habe ich nicht so viel Gesamtlast wie du
* BT - Disconnect Test: lief **200x** ohne crash durch, nachdem Zeile 72 im _BleHandler.cpp_ auf "`%i`" korrigiert wurde: `ESP_LOGI(TAG, "Disconnected Device %i", i);`

Ach ja: In der Git Version von _nimconfig.h_ ist das NimBLE Pinning noch auf "0" zurück gestellt^^

Ich denke mal, es kann nach den obigen Änderungen so in den Master Branch 👍

**Reply by shining-man** [reply]
Ich habe es jetzt wieder in dem main Branch gemerget.

**Comment by shining-man** [comment]
Heute hat meine Testbaugruppe, die jetzt einige Tage lief, einen Reboot gemacht und beim Neuanlauf hatte sie dann beim BT Connect immer wieder einen Rebbot gemacht, bis es irgendwann wieder lief. Die gleichen Probleme die ich damals hatte.
Kannst du einmal testen wie sich deine 7 NEEYs verhalten, wenn du sowohl den BT-Task als auch NimBLE auf Core 1 pinnst?
Aktuell ist beides auf Core 0 gepinnt.

**Reply by kmn13** [reply]
Auweia...

Hab jetzt mal schnell auf Core 1 umgepinnt und damit kommt ab >1 NEEY verbunden schon mal zyklisch die Meldung: `NimBLE: Controller not ready to receive packets (cmd)`
Mit jedem weitern NEEY steigt die Anzahl der Meldungen (kommen dann immer blockweise).

Jetzt nach ein paar Minuten haben die NEEYs begonnen sich zu trennen...ich kann leider noch nicht sagen warum. Aber eine Verbindung baut sich auch nicht mehr auf.
Log: 
```
D (456721) NimBLE: Controller not ready to receive packets (cmd)
D (457114) NimBLE: Controller not ready to receive packets (cmd)
D (459115) NimBLE: Controller not ready to receive packets (cmd)
I (461116) BLE_HANDLER: Disconnected: 3c:a5:51:89:43:aa
I (461117) BLE_HANDLER: Disconnected Device 5
D (461117) NimBLE: Controller not ready to receive packets (cmd)
D (461907) BLE_HANDLER: doStartBtScan -> Device 5
I (461907) BLE_HANDLER: Starte BT Scan
I (463120) BLE_HANDLER: Disconnected: 3c:a5:51:89:40:24
I (463121) BLE_HANDLER: Disconnected Device 3
D (463121) NimBLE: Controller not ready to receive packets (cmd)
I (465124) BLE_HANDLER: Disconnected: 3c:a5:51:89:35:27
I (465125) BLE_HANDLER: Disconnected Device 6
I (465125) BLE_HANDLER: Disconnected: 3c:a5:51:89:42:b2
I (465127) BLE_HANDLER: Disconnected Device 1
I (465132) BLE_HANDLER: Disconnected: 3c:a5:51:89:41:be
I (465136) BLE_HANDLER: Disconnected Device 0
I (465141) BLE_HANDLER: Disconnected: 3c:a5:51:94:40:a7
I (465145) BLE_HANDLER: Disconnected Device 4
I (465150) BLE_HANDLER: Disconnected: 3c:a5:51:89:41:cd
I (465154) BLE_HANDLER: Disconnected Device 2
CRIT NimBLEDevice: Resetting state; reason=17,
D (465163) NimBLE: Controller not ready to receive packets (cmd)
E NimBLEScan: Error initiating GAP discovery procedure; rc=21,
```
In diesem Stand betreibe ich bei nur die NEEYs.

**Reply by shining-man** [reply]
Ok, danke. Dann macht das Umpinnen keinen Sinn. Ich werde wohl das Problem finden müssen. Der Backtrace hilft mir auch nicht weiter, da CORRUPTED. 
Was war die längste Zeit die du die 7 NEEYS laufen gelassen hast und gab es dabei einmal einen Reboot?

**Reply by kmn13** [reply]
bei mir liefen die am Core 0 etwas mehr als 24h ohne Probleme oder Reboots.
Ich musste danach wieder etwas umbauen bzw meinen Laptop anderweitig nutzen, sodass ich stoppen musste.

**Reply by kmn13** [reply]
Ich hatte nun ein sepparates Testsystem aufgebaut, welches nur NEEY und Webserver übersicht offen spielt.
Nach 18h trennte sich 1x NEEY mit `disconnect, reason=520` baute aber gleich wieder eine Verbindung auf (müsste: BLE_ERR_CONN_SPVN_TMO - Connection Timeout sein).
Nun ein paar Stunden später ist alles abgeschmirrt...keine Ausgabe seriellen Log oder im Logfile ...hmm.... und per IP nicht mehr erreichbar.

Ich habe immer wieder einmal einen Blick auf den Heap geworfen, der war eigentlich immer annähernd gleich.

Muss das noch einmal wiederholen...

**Reply by shining-man** [reply]
Kann es auch sein, dass die NEEY Kommunikation weiterhin lief und nur der Webserver abgeschmiert ist? An der BSC HW siehst du das anhand der blinkenden LED. 
Wie hast du festgestellt, dass er abgeschmiert ist? Was war hierbei deine Aktion?

Du könntest in der main-Loop und in den Tasks ein Lebenszeichen, z.B. alle Sekunde, auf der seriellen-Schnittstelle ausgeben lassen, um zu sehen welcher Task abschmiert.

Wenn der Heap überlaufen würde, dann gebe es einen Reboot, den hattest du ja aber nicht...  Das wäre schöner, denn dann hätten wir einen Backtrace.

**Reply by kmn13** [reply]
Die "orig" BSC HW habe ich ja leider (nocht) nicht ;-)
Erkannt dadurch: alles NEEY getrennt (dann blinkt dort die blaue LED) aber keine Info im seriellen Trace (da kommt ja sonst der ganze Disconnect output). BSC auch nicht mehr erreichbar...auch nicht per Ping.

Hab mal in jede Task einen Hearbeat ins Log eingepflanzt.
